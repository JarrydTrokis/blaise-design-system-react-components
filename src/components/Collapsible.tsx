import React, {ReactElement, useState} from 'react';


export interface Props {
    children: ReactElement
    title: string
}

function Collapsible({children, title}: Props) {
    const [panelOpen, setPanelOpen] = useState<boolean>(false);

    return (
        <>
            <div id="collapsible-with-save"
                 className={`collapsible js-collapsible collapsible--initialised ${(panelOpen && "collapsible--open")} u-mt-l`}
                 data-btn-close="Hide this" data-save-state="true" role="group">
                <div className="collapsible__heading js-collapsible-heading"
                     role="link"
                     data-testid="collapsible-heading"
                     onClick={() => setPanelOpen(!panelOpen)}
                     onKeyPress={() => setPanelOpen(!panelOpen)}
                     tabIndex={0}
                     aria-expanded={panelOpen ? "true": "false"}
                     aria-controls="collapsible-with-save"
                     data-ga-action="Close panel">
                    <div className="collapsible__controls">
                        <div className="collapsible__title">{title}</div>
                        <span className="collapsible__icon">
                            <svg className="svg-icon"
                                 viewBox="0 0 7.5 12.85"
                                 xmlns="http://www.w3.org/2000/svg"
                                 focusable="false">
                                <path
                                    d="M5.74,14.28l-.57-.56a.5.5,0,0,1,0-.71h0l5-5-5-5a.5.5,0,0,1,0-.71h0l.57-.56a.5.5,0,0,1,.71,0h0l5.93,5.93a.5.5,0,0,1,0,.7L6.45,14.28a.5.5,0,0,1-.71,0Z"
                                    transform="translate(-5.02 -1.59)"/>
                            </svg>
                        </span>
                    </div>
                </div>
                <div id="collapsible-with-save-content" className="collapsible__content js-collapsible-content"
                     aria-hidden={(panelOpen ? "false" : "true")} data-testid="collapsible-content">

                    {children}

                    <button type="button" className="btn btn--small js-collapsible-button btn--secondary u-mt-m"
                            aria-hidden="true" aria-controls="collapsible-with-save" data-ga-action="Close panel"
                            onClick={() => setPanelOpen(false)}>
                        <span className="btn__inner js-collapsible-button-inner">Hide this</span>
                        <span className="btn__context u-vh">{title}</span>
                    </button>
                </div>
            </div>
        </>
    );
}

export default Collapsible;