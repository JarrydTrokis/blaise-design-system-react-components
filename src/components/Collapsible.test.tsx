import React from "react";
import {cleanup, fireEvent, render} from "@testing-library/react";
import Collapsible from "./Collapsible";

describe("Collapsible Test", () => {
    afterEach(() => {
        cleanup();
    });

    const contentText = "Inside the Collapsible"

    const Props = {
        title: "What is the Collapsible title?",
        content: <p>{contentText}</p>
    };

    it("matches Snapshot", () => {
        const wrapper = render(<Collapsible title={Props.title}>{Props.content}</Collapsible>);
        expect(wrapper).toMatchSnapshot();
    });

    it("renders tile", () => {
        const wrapper = render(<Collapsible title={Props.title}>{Props.content}</Collapsible>);
        expect(wrapper.queryAllByText(Props.title)).toHaveLength(2);
    });

    it("renders children inside the Collapsible", () => {
        const wrapper = render(<Collapsible title={Props.title}>{Props.content}</Collapsible>);
        expect(wrapper.getByText(contentText)).toBeInTheDocument();
    });

    it("children inside the Collapsible should be hidden by aria by default", () => {
        const wrapper = render(<Collapsible title={Props.title}>{Props.content}</Collapsible>);
        expect(wrapper.getByTestId("collapsible-content")).toHaveAttribute("aria-hidden", "true");
    });

    it("should change the to aria-hidden false when title is clicked on", () => {
        const wrapper = render(<Collapsible title={Props.title}>{Props.content}</Collapsible>);
        fireEvent.click(wrapper.getByRole("link"));
        expect(wrapper.getByTestId("collapsible-content")).toHaveAttribute("aria-hidden", "false");
    });

    it("should change the to aria-hidden false when you press the SpaceBar key on the title", () => {
        const wrapper = render(<Collapsible title={Props.title}>{Props.content}</Collapsible>);
        fireEvent.keyPress(wrapper.getByTestId("collapsible-heading"), { key: "Space", code: 32, charCode: 32 });
        expect(wrapper.getByTestId("collapsible-content")).toHaveAttribute("aria-hidden", "false");

    });

    it("should change the to aria-hidden false when you press the enter key on the title", () => {
        const wrapper = render(<Collapsible title={Props.title}>{Props.content}</Collapsible>);
        fireEvent.keyPress(wrapper.getByTestId("collapsible-heading"), { key: "Enter", code: 13, charCode: 13 });
        expect(wrapper.getByTestId("collapsible-content")).toHaveAttribute("aria-hidden", "false");

    });


    it("should change back to aria-hidden true when inner button 'Hide this' is pressed" , () => {
        const wrapper = render(<Collapsible title={Props.title}>{Props.content}</Collapsible>);
        fireEvent.click(wrapper.getByRole("link"));
        expect(wrapper.getByTestId("collapsible-content")).toHaveAttribute("aria-hidden", "false");


        fireEvent.click(wrapper.getByText("Hide this"));
        expect(wrapper.getByTestId("collapsible-content")).toHaveAttribute("aria-hidden", "true");
    });

});
