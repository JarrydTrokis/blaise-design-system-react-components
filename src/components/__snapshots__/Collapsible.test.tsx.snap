// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collapsible Test matches Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="collapsible js-collapsible collapsible--initialised false u-mt-l"
        data-btn-close="Hide this"
        data-save-state="true"
        id="collapsible-with-save"
        role="group"
      >
        <div
          aria-controls="collapsible-with-save"
          aria-expanded="false"
          class="collapsible__heading js-collapsible-heading"
          data-ga-action="Close panel"
          data-testid="collapsible-heading"
          role="link"
          tabindex="0"
        >
          <div
            class="collapsible__controls"
          >
            <div
              class="collapsible__title"
            >
              What is the Collapsible title?
            </div>
            <span
              class="collapsible__icon"
            >
              <svg
                class="svg-icon"
                focusable="false"
                viewBox="0 0 7.5 12.85"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5.74,14.28l-.57-.56a.5.5,0,0,1,0-.71h0l5-5-5-5a.5.5,0,0,1,0-.71h0l.57-.56a.5.5,0,0,1,.71,0h0l5.93,5.93a.5.5,0,0,1,0,.7L6.45,14.28a.5.5,0,0,1-.71,0Z"
                  transform="translate(-5.02 -1.59)"
                />
              </svg>
            </span>
          </div>
        </div>
        <div
          aria-hidden="true"
          class="collapsible__content js-collapsible-content"
          data-testid="collapsible-content"
          id="collapsible-with-save-content"
        >
          <p>
            Inside the Collapsible
          </p>
          <button
            aria-controls="collapsible-with-save"
            aria-hidden="true"
            class="btn btn--small js-collapsible-button btn--secondary u-mt-m"
            data-ga-action="Close panel"
            type="button"
          >
            <span
              class="btn__inner js-collapsible-button-inner"
            >
              Hide this
            </span>
            <span
              class="btn__context u-vh"
            >
              What is the Collapsible title?
            </span>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="collapsible js-collapsible collapsible--initialised false u-mt-l"
      data-btn-close="Hide this"
      data-save-state="true"
      id="collapsible-with-save"
      role="group"
    >
      <div
        aria-controls="collapsible-with-save"
        aria-expanded="false"
        class="collapsible__heading js-collapsible-heading"
        data-ga-action="Close panel"
        data-testid="collapsible-heading"
        role="link"
        tabindex="0"
      >
        <div
          class="collapsible__controls"
        >
          <div
            class="collapsible__title"
          >
            What is the Collapsible title?
          </div>
          <span
            class="collapsible__icon"
          >
            <svg
              class="svg-icon"
              focusable="false"
              viewBox="0 0 7.5 12.85"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5.74,14.28l-.57-.56a.5.5,0,0,1,0-.71h0l5-5-5-5a.5.5,0,0,1,0-.71h0l.57-.56a.5.5,0,0,1,.71,0h0l5.93,5.93a.5.5,0,0,1,0,.7L6.45,14.28a.5.5,0,0,1-.71,0Z"
                transform="translate(-5.02 -1.59)"
              />
            </svg>
          </span>
        </div>
      </div>
      <div
        aria-hidden="true"
        class="collapsible__content js-collapsible-content"
        data-testid="collapsible-content"
        id="collapsible-with-save-content"
      >
        <p>
          Inside the Collapsible
        </p>
        <button
          aria-controls="collapsible-with-save"
          aria-hidden="true"
          class="btn btn--small js-collapsible-button btn--secondary u-mt-m"
          data-ga-action="Close panel"
          type="button"
        >
          <span
            class="btn__inner js-collapsible-button-inner"
          >
            Hide this
          </span>
          <span
            class="btn__context u-vh"
          >
            What is the Collapsible title?
          </span>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
